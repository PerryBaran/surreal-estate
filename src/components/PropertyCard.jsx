import React from "react";
import PropTypes from "prop-types";
import style from "../styles/propertyCard.module.css";
import { bath, bed, email as emailIcon } from "../media/icons";

const PropertyCard = ({
  title,
  type,
  bathrooms,
  bedrooms,
  price,
  city,
  email,
}) => {
  return (
    <div className={style["property-card"]}>
      <h2 className={style["property-card-title"]}>{title}</h2>
      <p className={style["property-card-info"]}>
        {type} - {city}
      </p>
      <div className={style["property-card-rooms"]}>
        <div className={style["property-card-room"]}>
          <img src={bath} alt="bathrooms" />
          <p>{bathrooms}</p>
        </div>
        <div className={style["property-card-room"]}>
          <img src={bed} alt="bedrooms" />
          <p>{bedrooms}</p>
        </div>
      </div>
      <p className={style["property-card-price"]}>Â£ {price.toFixed(2)}</p>
      <a href={`mailto:${email}`} className={style["property-card-email"]}>
        <img src={emailIcon} alt="email" />
        <p>Email</p>
      </a>
    </div>
  );
};

PropertyCard.propTypes = {
  title: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  bathrooms: PropTypes.number.isRequired,
  bedrooms: PropTypes.number.isRequired,
  price: PropTypes.number.isRequired,
  city: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
};

export default PropertyCard;
