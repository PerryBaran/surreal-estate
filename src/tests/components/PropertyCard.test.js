import React from "react";
import { render, screen } from "@testing-library/react";
import PropertyCard from "../../components/PropertyCard";

describe("PropertyCard", () => {
  describe("falsy userId", () => {
    const validProps = {
      _id: "4refsf",
      title: "2 bed flat",
      type: "flat",
      bathrooms: "1",
      bedrooms: "2",
      price: "5000",
      city: "Leeds",
      email: "not.real@email.com",
      userId: "",
      onSaveProperty: jest.fn(),
      onRemoveProperty: jest.fn(),
    };

    test("snapshot", () => {
      const { asFragment } = render(<PropertyCard {...validProps} />);

      expect(asFragment()).toMatchSnapshot();
    });

    test("renders correctly", () => {
      render(<PropertyCard {...validProps} />);
      const bathroomsImg = screen.getByAltText("bathrooms");
      const bedroomsImg = screen.getByAltText("bedrooms");
      const emailImg = screen.getByAltText("email");

      expect(screen.getByAltText("property")).toHaveAttribute(
        "src",
        "card.png"
      );
      expect(screen.getByText(validProps.title)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getByText(new RegExp(validProps.city))).toHaveTextContent(
        `${validProps.type} - ${validProps.city}`
      );
      expect(bathroomsImg).toHaveAttribute("alt", "bathrooms");
      expect(bathroomsImg).toHaveAttribute("src", "bath.png");
      expect(screen.getByText(validProps.bathrooms)).toBeInTheDocument();
      expect(bedroomsImg).toHaveAttribute("alt", "bedrooms");
      expect(bedroomsImg).toHaveAttribute("src", "bed.png");
      expect(screen.getByText(validProps.bedrooms)).toBeInTheDocument();
      expect(screen.getByText(new RegExp(validProps.price))).toHaveTextContent(
        `Â£ ${Number(validProps.price).toFixed(2)}`
      );
      expect(screen.getByText(/email/i).closest("a")).toHaveAttribute(
        "href",
        `mailto:${validProps.email}`
      );
      expect(emailImg).toHaveAttribute("src", "email.png");
      expect(emailImg.closest("a")).toHaveTextContent(/email/i);
    });
  });

  describe("truthy userId", () => {
    const validProps = {
      _id: "4refsf",
      title: "2 bed flat",
      type: "flat",
      bathrooms: "1",
      bedrooms: "2",
      price: "5000",
      city: "Leeds",
      email: "not.real@email.com",
      userId: "truthy",
      onSaveProperty: jest.fn(),
      onRemoveProperty: jest.fn(),
    };

    test("snapshot", () => {
      const { asFragment } = render(<PropertyCard {...validProps} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("truthy userId and favouriteId", () => {
    const validProps = {
      _id: "4refsf",
      title: "2 bed flat",
      type: "flat",
      bathrooms: "1",
      bedrooms: "2",
      price: "5000",
      city: "Leeds",
      email: "not.real@email.com",
      userId: "truthy",
      favouriteId: "alsoTruthy",
      onSaveProperty: jest.fn(),
      onRemoveProperty: jest.fn(),
    };

    test("snapshot", () => {
      const { asFragment } = render(<PropertyCard {...validProps} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
