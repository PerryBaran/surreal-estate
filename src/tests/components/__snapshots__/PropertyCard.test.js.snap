// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyCard falsy userId snapshot 1`] = `
<DocumentFragment>
  <div
    class="property-card"
  >
    <img
      alt="property"
      class="property-card__image"
      src="card.png"
    />
    <h2>
      2 bed flat
    </h2>
    <p>
      flat - Leeds
    </p>
    <div
      class="property-card__rooms"
    >
      <div
        class="property-card__rooms__room"
      >
        <img
          alt="bathrooms"
          class="property-card__icon"
          src="bath.png"
        />
        <p>
          1
        </p>
      </div>
      <div
        class="property-card__rooms__room"
      >
        <img
          alt="bedrooms"
          class="property-card__icon"
          src="bed.png"
        />
        <p>
          2
        </p>
      </div>
    </div>
    <p
      class="property-card__price"
    >
      £ 5000.00
    </p>
    <a
      class="property-card__email"
      href="mailto:not.real@email.com"
    >
      <img
        alt="email"
        class="property-card__icon"
        src="email.png"
      />
      <p>
        Email
      </p>
    </a>
  </div>
</DocumentFragment>
`;

exports[`PropertyCard truthy userId and favouriteId snapshot 1`] = `
<DocumentFragment>
  <div
    class="property-card"
  >
    <img
      alt="property"
      class="property-card__image"
      src="card.png"
    />
    <h2>
      2 bed flat
    </h2>
    <p>
      flat - Leeds
    </p>
    <div
      class="property-card__rooms"
    >
      <div
        class="property-card__rooms__room"
      >
        <img
          alt="bathrooms"
          class="property-card__icon"
          src="bath.png"
        />
        <p>
          1
        </p>
      </div>
      <div
        class="property-card__rooms__room"
      >
        <img
          alt="bedrooms"
          class="property-card__icon"
          src="bed.png"
        />
        <p>
          2
        </p>
      </div>
    </div>
    <p
      class="property-card__price"
    >
      £ 5000.00
    </p>
    <a
      class="property-card__email"
      href="mailto:not.real@email.com"
    >
      <img
        alt="email"
        class="property-card__icon"
        src="email.png"
      />
      <p>
        Email
      </p>
    </a>
    <button
      class="remove-property"
      type="button"
    >
      <span>
        Saved
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`PropertyCard truthy userId snapshot 1`] = `
<DocumentFragment>
  <div
    class="property-card"
  >
    <img
      alt="property"
      class="property-card__image"
      src="card.png"
    />
    <h2>
      2 bed flat
    </h2>
    <p>
      flat - Leeds
    </p>
    <div
      class="property-card__rooms"
    >
      <div
        class="property-card__rooms__room"
      >
        <img
          alt="bathrooms"
          class="property-card__icon"
          src="bath.png"
        />
        <p>
          1
        </p>
      </div>
      <div
        class="property-card__rooms__room"
      >
        <img
          alt="bedrooms"
          class="property-card__icon"
          src="bed.png"
        />
        <p>
          2
        </p>
      </div>
    </div>
    <p
      class="property-card__price"
    >
      £ 5000.00
    </p>
    <a
      class="property-card__email"
      href="mailto:not.real@email.com"
    >
      <img
        alt="email"
        class="property-card__icon"
        src="email.png"
      />
      <p>
        Email
      </p>
    </a>
    <button
      class="save-property"
      type="button"
    >
      Save
    </button>
  </div>
</DocumentFragment>
`;
